src/cachetools/__init__.py

```diff
@@ -526,20 +527,32 @@ def wrapper(*args, **kwargs):
 
         else:
 
+            _key_level_locks = {}
+
             def wrapper(*args, **kwargs):
                 k = key(*args, **kwargs)
-                try:
-                    with lock:
+                with lock:
+                    try:
                         return cache[k]
-                except KeyError:
-                    pass  # key not found
-                v = func(*args, **kwargs)
-                # in case of a race, prefer the item already in the cache
-                try:
-                    with lock:
-                        return cache.setdefault(k, v)
-                except ValueError:
-                    return v  # value too large
+                    except KeyError:
+                        pass  # key not found
+                    _key_level_lock = _key_level_locks.setdefault(k, RLock())
+                # Only one caller may generate or retrieve a value for this key
+                with _key_level_lock:
+                    try:
+                        with lock:
+                            # In the case that this thread was blocked, retrieve and return the now computed value
+                            return cache[k]
+                    except KeyError:
+                        # Otherwise compute it on this thread since the key is not in the cache
+                        v = func(*args, **kwargs)
+                    finally:
+                        _key_level_locks.pop(k)
```

Comment:
Unlocked access to _key_level_locks - since `k` provides it's own `__hash__` function, `dict` operations are not thread-safe, not even in CPython with its global interpreter lock...

Reply:
It does not look like the GIL get's released, however I've protected this call either way since it doesn't look like it's going to give rise to a deadlock (all operations under the global (non key level) lock are context managed and finite)

https://github.com/python/cpython/blob/main/Objects/setobject.c#L630